### PR TITLE
Support OutputDebugString

### PR DESCRIPTION
**Pull request by JayFoxRox**
Very useful function for getting data to a debugger.

### CODE REVIEW

File: lib/xboxkrnl/xboxkrnl.h [hunk 1]
```diff
@@ -9,6 +9,7 @@
 #pragma once
 
 #include <xboxkrnl/xboxdef.h>
+#include <xboxkrnl/xboxnt.h>
```
Comment (JayFoxRox) [inline lib/xboxkrnl/xboxkrnl.h]: Reminder: These will probably need this in the future:

```
/* stop clang from crying */
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wlanguage-extension-token"

/* MSVC-compatibility for structure-packing */
#pragma ms_struct on
```

*(Can be done after this PR, but it's something worth noting before we add even more)*

File: lib/winapi/winbase.h [hunk 2]
```diff
@@ -2,6 +2,7 @@
 #define __WINBASE_H__
 
 #include <windef.h>
+#include <winnt.h>
```
Comment (JayFoxRox) [inline lib/winapi/winbase.h]: I'm not sure if this should be included from here, or from windef or elsewhere.

I don't remember why I put this here.

File: lib/xboxkrnl/xboxkrnl.h [hunk 3]
```diff
@@ -132,12 +133,12 @@ typedef enum _MODE {
 /**
  * This struct defines a counted string used for ANSI-strings
  */
-typedef struct _ANSI_STRING
+typedef struct
```
Comment (dracc) [inline lib/xboxkrnl/xboxkrnl.h]: Anonymous structs is a GCC extension, do we want to use those or should we name structs when we have the opportunity?

Comment (JayFoxRox) [reply lib/xboxkrnl/xboxkrnl.h]: This change also causes `warning: incompatible pointer types passing 'POBJECT_STRING'
      (aka 'STRING *') to parameter of type 'PANSI_STRING' (aka 'struct _STRING *') [-Wincompatible-pointer-types]`

So I'll have to look into fixing this in general.

Comment (JayFoxRox) [reply lib/xboxkrnl/xboxkrnl.h]: Fixed by reusing the `STRING` type.

File: lib/xboxkrnl/xboxkrnl.h [hunk 4]
```diff
@@ -132,12 +132,7 @@ typedef enum _MODE {
 /**
  * This struct defines a counted string used for ANSI-strings
  */
-typedef struct _ANSI_STRING
-{
-    USHORT Length;        /**< Length in bytes of the string stored in the buffer */
-    USHORT MaximumLength; /**< Length in bytes of the buffer */
-    PSTR Buffer;          /**< Pointer to the buffer used for the character-string */
-} ANSI_STRING, *PANSI_STRING, STRING, *PSTRING, OBJECT_STRING, *POBJECT_STRING;
+typedef STRING OBJECT_STRING, *POBJECT_STRING;
```
Comment (JayFoxRox) [inline lib/xboxkrnl/xboxkrnl.h]: The comment for this is broken.

File: lib/winapi/debug.c [hunk 5]
```diff
@@ -0,0 +1,17 @@
+#include <winbase.h>
+#include <winnt.h>
+
+#include <string.h>
+
+void WINAPI OutputDebugStringA (LPCTSTR lpOutputString)
```
Comment (thrimbor) [inline lib/winapi/debug.c]: MS docs describe that this function calls `DbgPrint()` internally. I couldn't spot any difference in behavior when using this functions vs. using DbgPrint (when called on Retail or Debug BIOSes), so is this something we should do, too?

File: lib/winapi/winbase.h [hunk 6]
```diff
@@ -11,6 +12,14 @@ extern "C"
 DWORD GetLastError (void);
 void SetLastError (DWORD error);
 
+void WINAPI OutputDebugStringA (LPCTSTR lpOutputString);
```
Comment (thrimbor) [inline lib/winapi/winbase.h]: [MS docs](https://docs.microsoft.com/en-us/windows/desktop/api/debugapi/nf-debugapi-outputdebugstringa) indicate that this should be located in a header called `debugapi.h`.

File: lib/xboxkrnl/xboxkrnl.h (hunk 4 above)
Comment (thrimbor) [reply lib/xboxkrnl/xboxkrnl.h]: Should that typedef get moved, too?

File: lib/winapi/debug.c (hunk 5 above)
Comment (JayFoxRox) [reply lib/winapi/debug.c]: MS docs are either wrong (changes over time?), or `DbgPrint` was inlined (which I doubt).

My version is based on reverse engineering of games, and it does not call any function which does string formatting (which `DbgPrint` would do).

Another difference is that this one doesn't need a function call and can be LTCG'd; it also isn't imported from the kernel, which can be relevant in some cases.

File: lib/winapi/winbase.h (hunk 6 above)
Comment (JayFoxRox) [reply lib/winapi/winbase.h]: I believe there's 2 different versions of this function: for userspace, and for kernel drivers.

https://source.winehq.org/WineAPI/OutputDebugStringA.html (and other languages of the MSDN article) claim it's in winbase.h, and I believe I found the location from a copy of an old DDK myself.

It's also possible that they moved the function over time to make winbase.h / windows.h more modular.

File: lib/xboxkrnl/xboxkrnl.h (hunk 4 above)
Comment (JayFoxRox) [reply lib/xboxkrnl/xboxkrnl.h]: The object API is part of the kernel; I'm not sure if anything else uses it.

I think this is a custom name for Xbox anyway - I believe I checked Windows SDK and DDK and wasn't able to find this name? Current MSDN documents the users as `UNICODE_STRING` (which would be equivalent to a generic `ANSI_STRING`?) as far as I know.

Also I had no use for it as part of `OutputDebugString`, so I didn't want to touch it.

**Comment by thrimbor** [comment]
Code appears to be fine and nobody voiced any concerns, so if you resolve the conflicts I'll merge later today.

**Comment by thrimbor** [comment]
Turns out I can force push to PR-ed branches, so I took the liberty of rebasing this myself. Merging now.

**Pull request merged**